File: android/app/src/main/java/io/wantic/app/share/models/Wish.kt
package io.wantic.app.share.models

import io.wantic.app.share.utils.UUIDSerializer
import kotlinx.serialization.Serializable
import java.util.*

@Serializable
class Wish constructor(
    var id: @Serializable(with = UUIDSerializer::class) UUID?,
    val wishListId: @Serializable(with = UUIDSerializer::class) UUID,
    val name: String, val price: Price, val note: String?,
    val productUrl: String,
    val imageUrl: String?,
    val isFavorite: Boolean
    ) {

    companion object {
        fun create(productInfo: ProductInfo, wishListId: UUID): Wish {
            val price = Price(productInfo.price)
            return Wish(
                null,
                wishListId,
                productInfo.name,
                price,
                productInfo.note,
                productInfo.productUrl,
                productInfo.imageUrl,
                productInfo.isFavorite
            )
        }
    }

}

File: ios/App/Share/Model/Wish/Wish.swift
//
//  Wish.swift
//  Share
//
//  Created by Alex on 14/11/2023.
//

import Foundation

struct Wish: Codable {
    
    var id: UUID?
    var wishListId: UUID?
    var name: String?
    var note: String?
    var price: Price
    var coupon: Coupon?
    var productUrl: String?
    var imageUrl: String?
    var isFavorite: Bool = false
    var isCreator: Bool? = false
    
    var isValid: Bool {
        wishListId != nil && name != nil && productUrl != nil
    }
    
    init() {
        self.price = Price(amount: 0.00)
    }
    
    init(_ webPageInfo: WebPageInfo, webPageImage: WebPageImage?, isCreator: Bool? = false) {
        
        if let webPageImageName = webPageImage?.name, !webPageImageName.isEmpty {
            name = webPageImageName
        } else {
            name = webPageInfo.title
        }
        name?.truncateIfNeeded()

        productUrl = webPageInfo.url
        imageUrl = webPageImage?.url
        price = webPageInfo.price
        coupon = webPageInfo.coupon
        self.isCreator = isCreator
    }
}
